### PR TITLE
Compress the data when saving annotions

### PR DESCRIPTION
**Pull request by calixteman**
CompressionStream API has been added in Firefox 113
(see https://bugzilla.mozilla.org/show_bug.cgi?id=1823619)
hence we can use it to compress the streams with added/modified annotations.

### CODE REVIEW

File: src/core/writer.js
Comment (Snuffleupagus) [inline src/core/writer.js:56]: Given that we might not actually be able to compress, if the functionality isn't available in the browser, how about calling it `shouldCompress` instead (which feels slightly less absolute)?

File: src/core/writer.js
Comment (Snuffleupagus) [inline src/core/writer.js:82]: Please note that the only *generally safe* way to feature-test for a global property, which might not be available, is to use a `typeof`-check. 

That being said, based on a quick look at a few MDN articles it really seems to me a bunch of the functionality used (indirectly) below may not be fully available in all browsers that we currently support and that it'd probably be best to just wrap all of this in try-catch to avoid bugs.
```suggestion
  if (mustCompress && typeof CompressionStream !== "undefined") {
    try {
      const byteArray = stringToBytes(string);
      // eslint-disable-next-line no-undef
      const cs = new CompressionStream("deflate");
      const writer = cs.writable.getWriter();
      writer.write(byteArray);
      writer.close();
      const compressedData = await new Response(cs.readable).arrayBuffer();
      string = bytesToString(new Uint8Array(compressedData));
    } catch (ex) {
      info(`writeStream - cannot compress data: "${ex}".`);
    }
  }
```

File: src/core/writer.js
Comment (Snuffleupagus) [inline src/core/writer.js:81]: I'm assuming that the fact that it's documented as returning UTF-8 encoded data, according to https://developer.mozilla.org/en-US/docs/Web/API/Response/text, means that we can't simply do the following?
```suggestion
    string = await new Response(cs.readable).text();
```

**Review by Snuffleupagus — COMMENTED** [review]
I'd probably suggest that we don't land this just before a general PDF.js library release, given that the patch uses a bunch of "newer" functionality.

File: src/core/writer.js
Comment (Snuffleupagus) [inline src/core/writer.js:55]: It wouldn't hurt with a (short) comment explaining how this value was chosen :-)

File: src/core/writer.js
Comment (Snuffleupagus) [inline src/core/writer.js:67]: There's a subtle problem with this check, since it doesn't actually take the `Name`-caching (in `src/core/primitives.js)` fully into account. Hence the following is probably necessary to be guaranteed to avoid any problems.
```suggestion
    } else if (!(filter instanceof Name) || filter.name !== flateDecode.name) {
```

File: src/core/writer.js
Comment (Snuffleupagus) [inline src/core/writer.js:62]: Given that the actual compression can actually fail below, e.g. if the necessary functionality isn't available, you probably shouldn't update the filter entry until *afterwards* to prevent bugs!?

File: src/core/writer.js
Comment (Snuffleupagus) [inline src/core/writer.js:69]: Given that the actual compression can fail below, e.g. if the necessary functionality isn't available, you probably shouldn't update the filter entry until *afterwards* to prevent bugs!?

File: test/unit/writer_spec.js
Comment (Snuffleupagus) [inline test/unit/writer_spec.js:1]: Can we please add at least one unit-test that checks that compression works as expected?

File: src/core/writer.js
Comment (Snuffleupagus) [inline src/core/writer.js:72]: Given the `Name` caching, please change this into the following to prevent any future bugs.
```suggestion
    (filter instanceof Name && filter.name === flateDecode.name)
```

**Review by Snuffleupagus — APPROVED** [review]
r=me, with one more comment; thank you!

As mentioned previously, I'd suggest that we wait until *after* the next release (hopefully tomorrow) before landing this.

File: src/core/writer.js
Comment (calixteman) [reply src/core/writer.js:72]: I thought that the cache was useful to have only one instance of a given Name, e.g. `Name.get("FlateDecode")` is used for whatever `/FlatDecode` we've in the pdf and in such a case `filter===flatdecode` should be enough.
And if my understanding is wrong, what's the point to make a lookup in a hashmap when we could just use a `new Name...` ?

Comment (Snuffleupagus) [reply src/core/writer.js:72]: The caching does make sense, and will work in *most* cases. However please keep in mind that (all) caches can be cleared, e.g. during idle cleanup, and in that case it's no longer guaranteed that e.g. `Name`-objects match.

Consider the following order of events:
 - The stream-dict /Filter entry is read and initialized somewhere else in the code-base, thus creating a new `Name`-instance and adding it to the `NameCache`. Furthermore, the dict-property is also being cached on the `XRef`-instance.
 - Some time passes, currently 30 seconds of inactivity in the viewer, which cause the `NameCache` to be cleared.
 - The user tries to save a modified document, and here the stream-dict /Filter entry now points to an *old* `Name`-instance (since the dict is cached on the `XRef`-instance). However, the `Name.get("FlateDecode");` call above creates a *new* `Name`-instance in the `NameCache` and we thus get a mismatch unless we compare the `name`-properties.

Obviously this situation will be really rare, but if/when it does happens we'll have a very difficult to find and highly intermittent bug on our hands.

Comment (calixteman) [reply src/core/writer.js:72]: Ok, it makes sense.
What do you think of adding `Name.isEqual(n1, n2)` in Name with a comment (a copy/paste of the above comment will be enough I guess) explaining the reason for having it ?
It'll avoid to have to remember that (and to avoid the bus factor).

Comment (Snuffleupagus) [reply src/core/writer.js:72]: Sure, that could work (although the comment should probably be generalized slightly compared).
Maybe just following the same pattern as in
https://github.com/mozilla/pdf.js/blob/4d3dfe254ffe90afff336ec28024e951b15ffad0/src/core/primitives.js#L372-L380
and 
https://github.com/mozilla/pdf.js/blob/4d3dfe254ffe90afff336ec28024e951b15ffad0/test/unit/primitives_spec.js#L579-L591

Comment (calixteman) [reply src/core/writer.js]: Ok. Let's do that in a followup, would you mind to do it ?

Comment (Snuffleupagus) [reply src/core/writer.js]: Sure, I'll try to look into that later.

File: src/core/writer.js
Comment (Snuffleupagus) [inline src/core/writer.js:61]: Could it potentially be the case that the /Filter-entry is ever an indirect object, i.e. a `Ref`, in which case the code below probably won't work correctly?

Shouldn't we actually do the following instead...
```suggestion
  const filter = stream.dict.get("Filter");
```
... or in case you're worried about the data not being immediately available
```suggestion
  const filter = await stream.dict.getAsync("Filter");
```

**Comment by calixteman** [comment]
@Snuffleupagus, I had to revert the `Response.text()` thing because it wasn't working correctly finally (which makes sense since `text` is supposed to decode utf-8 and here we just have binary data).

**Comment by Snuffleupagus** [comment]
> I had to revert the `Response.text()` thing because it wasn't working correctly finally (which makes sense since `text` is supposed to decode utf-8 and here we just have binary data).

That doesn't really surprise me, and the added comment should prevent someone from trying to change that.

---

Given the number of Linux-bot movement I've not actually looked at the results, however the Windows-bot "failures" look like normal fallout from a Firefox update (to version 115 in this case); hence this seems fine to just land and makeref.

**Pull request merged**
